apps/storefront/src/pages/Login/index.tsx

```diff
@@ -131,51 +132,52 @@ export default function Login(props: PageProps) {
     logo: displayStoreLogo ? logo : undefined,
   };
 
-  useEffect(() => {
-    const init = async () => {
-      try {
-        const loginFlag = searchParams.get('loginFlag');
-        const showTipInfo = searchParams.get('showTip') !== 'false';
-
-        setShowTipInfo(showTipInfo);
+  const logoutEffect = useB2BCallback(B2BEvent.OnLogout, async (dispatchLogoutEvent) => {
+    try {
+      const loginFlag = searchParams.get('loginFlag');
+      const showTipInfo = searchParams.get('showTip') !== 'false';
 
-        if (loginFlag) setLoginFlag(loginFlag);
+      setShowTipInfo(showTipInfo);
 
-        if (loginFlag === '7') {
-          snackbar.error(b3Lang('login.loginText.invoiceErrorTip'));
-        }
-        if (loginFlag === '3' && isLoggedIn) {
-          const cartInfo = await getCart();
-
-          if (cartInfo.data.site.cart?.entityId) {
-            const deleteQuery = deleteCartData(cartInfo.data.site.cart.entityId);
-            await deleteCart(deleteQuery);
-          }
+      if (loginFlag) setLoginFlag(loginFlag);
 
-          const { result } = (await bcLogoutLogin()).data.logout;
+      if (loginFlag === '7') {
+        snackbar.error(b3Lang('login.loginText.invoiceErrorTip'));
+      }
+      if (loginFlag === '3' && isLoggedIn) {
+        const cartInfo = await getCart();
 
-          if (result !== 'success') return;
+        if (cartInfo.data.site.cart?.entityId) {
+          const deleteQuery = deleteCartData(cartInfo.data.site.cart.entityId);
+          await deleteCart(deleteQuery);
+        }
 
-          if (isMasquerade) {
-            await endMasquerade();
-          }
+        const { result } = (await bcLogoutLogin()).data.logout;
 
-          // SUP-1282 Clear sessionStorage to allow visitors to display the checkout page
-          window.sessionStorage.clear();
+        if (result !== 'success') return;
 
-          logoutSession();
-          setLoading(false);
-          return;
+        if (isMasquerade) {
+          await endMasquerade();
         }
 
+        // SUP-1282 Clear sessionStorage to allow visitors to display the checkout page
+        window.sessionStorage.clear();
+
+        logoutSession();
         setLoading(false);
-      } finally {
-        setLoading(false);
+        return;
       }
-    };
 
-    init();
-  }, [b3Lang, endMasquerade, isLoggedIn, isMasquerade, searchParams]);
+      setLoading(false);
+    } finally {
+      setLoading(false);
+      dispatchLogoutEvent();
+    }
+  });
+
+  useEffect(() => {
+    logoutEffect();
+  }, [logoutEffect]);
```

Comment:
This one I am not sure is smart.

`useB2BCallback` does not memoize itself (it will regenerate the function on every render).

and now we are running this effect logic every render essentially.
This is a big departure from previous logic, which only ran this effect when these values changed:

`[b3Lang, endMasquerade, isLoggedIn, isMasquerade, searchParams]`.

@bc-victor - Am I missing something here? Shouldn't this logoutEffect function be created within the useEffect like it was being done before to make sure we aren't introducing some unexpected new behaviour?

Reply:
Yeah that is a good observation - somehow React is not regenerating on every render which makes me think that the runtime is doing some memoization for us although not sure if it will work on every case.

@deov31 do you think you can follow Micah's suggestion and put the function back in the useEffect?

Reply:
@bc-victor 👍